### ISSUE TITLE
Body Alert Mail

### ISSUE THREAD

**Issue by fran-nlt** [librenms/librenms#1813]
Hi
I would like to customize the body of Alert Mail or add some lines to this message
![1](https://cloud.githubusercontent.com/assets/13962782/9586873/cd067b98-4ff5-11e5-8037-1da32cb51bb1.PNG)

**Comment by laf** [comment]
http://docs.librenms.org/Extensions/Alerting/#templates

This gives you the information you should need to build a custom template.

**Comment by fran-nlt** [comment]
This is apply in the Gui or Cli?

**Comment by laf** [comment]
WebUI. It's in Overview -> Alerts -> Templates

**Comment by fran-nlt** [comment]
Sorry but i don´t understading this page.

**Comment by laf** [comment]
It's not that difficult once you start, click around and see how the template gets built up. Use the example one in the docs to see a working version.

You can just enter the entire template into the box on the left from the docs and make the changes you might need.

**Comment by fran-nlt** [comment]
Hi!
I dont resolv this issue.
I dont can modify this body message.

Do you send capture how to?

**Comment by f0o** [comment]
Please read the docs, http://docs.librenms.org/Extensions/Alerting/#templates

Like @laf said, browse to the Alert-Templates page under `Overview -> Alerts -> Templates` and create your template, then assign it to your rule or rules.

**Comment by digitalpilotnm** [comment]
I would state that the instructions on templates is very UNCLEAR to those that are not actual developers on this code base.  While it may be clear to LAF and f0o, its not clear to others.   This is a classic case of the developers being blinded by being to close to the knowledge / code base.  I've spent several hours trying to get alert templates to be more informative, but have failed.

I want alerts that actually provide usable data.  

For example:  if a BGP session goes down, the ALERT should actually contain the ASN and the peer description / name, in the alert.  Something like:

Device:  cr01.lax.example.com
ASN:  226
ASN Name: AS226-LOS-NETTOS-LAX-ANY2IX
PeerIP: <IP of the PEER>
Timestamp:  <Some Time>
State:  {UP | DOWN }

Instead we get this
Device: cr02.lax.example.com
\r\n
Title: Op_Alert_Title
\r\n
elapsed: 4m 36s
\r\n
id:
\r\n
uid: 25154
Faults:\r\n
  #1: bgpPeer_id => 120; bgpPeerIdentifier => 1.2.3.4; \r\n
state: 1
severity: critical
rule: %bgpPeers.bgpPeerFsmEstablishedTime < "300" && %bgpPeers.bgpPeerState
= "established"
timestamp: 2016-04-09 21:50:25
contacts: Array
key: 1
value: Array

Which does no good in telling a NOC tech what is going on.  The NOC Tech needs to connect to the router, and hunt for BGP session 1.2.3.4.   When you have 500+ peers that's not quick / fun

So I would strongly suggest that more real world examples be provided by the development team so that us mere mortal users can better understand how things work.  Otherwise this is just Adam all over again.

**Comment by laf** [comment]
Show us you tried example and we can tell you where you are going wrong.

Many people have configured there own custom templates perfectly fine, some even going as far as linking graphs directly into the emails so things like traffic graphs can be seen. If those people haven't contributed examples back then we can't force them but to expect us to provide examples for different scenarios is above what we have time to provide.

**Comment by digitalpilotnm** [comment]
Hi,  I think it would be reasonable to have a few more examples that show
how things work.
Currently the docs contain a single example.  Maybe examples from your own
production systems,
since it would mostly be a cut/paste would be useful.
If I can get the gist of how to make these better, I'm happy to contribute
back.
Here is my current TEST template.  I'm unable to see the data I would like
to see on an alert
relating to a BGP issue as previously explained in this ticket.

Device: %hostname
\r\n
Title: %title
\r\n
elapsed: %elapsed
\r\n
id: %id
\r\n
uid: %uid
{if %faults}Faults:\r\n
{foreach %faults}  #%key: %value.string\r\n{/foreach}{/if}
state: %state
severity: %severity
rule: %rule
timestamp: %timestamp
contacts: %contacts
key: %key
value: %value

**Comment by davidlanline** [comment]
I agree with digitalpilotnm there really should be more examples, if this project is going to be truly useful, then useful alerts are an absolute necessity and creating those alert templates is somewhat obscure. The documentation in this area is very poor. I am a systems administrator, not a developer, if I have to be a developer to use your code, why would I use your code? At some point it becomes more time effective for me considering I only need 10-20% of this to build it my own way.  ::endrant::

I have the following template for device up/downs and I got most of it to work, however, I would like to add the content from devices.notes and devices.purpose to the email alert and I simply cannot figure out how to get it to work. 

Any help would be greatly appreciated. 


<i>%hostname </i><br />
Severity: %severity<br />


{if %faults}<b>Faults:</b><br />
{foreach %faults}<br />
</i>
<b>Reason Down:</b>%value.status_reason<br />
<b>Last Time Up:</b> %value.last_ping<br />
<b>Time Elapsed:</b> %elapsed <br />
<b>Device Description:</b>%value.sysDescr<br />
<b>Device Location:</b> %value.location<br />
<b>Device Purpose:</b> %value.purpose<br />
<b>Device Notes:</b>%value.notes<br />
Name:%value.sysName<br />
OS: %value.os<br />
{/foreach}<br />
{/if}

**Comment by laf** [comment]
To be clear here, this software is an open source project. No one gets paid to work on it, contribute to it, write docs for it or provide any form of help. If the documentation isn't good enough for you then feel free to expand on it and contribute back - we've made it extremely easy to be able to do this straight from the docs themselves. Above all, I think you will find this project is already truly useful to a large number of people.

As for your template, I've just tested the exact thing you've pasted here and it works for me:

```bash
localhost

Severity: critical

Faults:




Reason Down:snmp

Last Time Up: 2016-09-25 21:56:52

Time Elapsed: 3m 43s

Device Description:Linux web01.novalocal 3.10.0-327.4.5.el7.x86_64 #1 SMP
Mon Jan 25 22:07:14 UTC 2016 x86_64

Device Location: Rack, Room, Building, City, Country [GPSX,Y]

Device Purpose: boom boom

Device Notes:test 123

Name:web01.novalocal

OS: linux
```

You did apply the template to the rule didn't you?

**Comment by digitalpilotnm** [comment]
Hi,   Personally the "hey its open source" response only goes so far.  If
the developers aren't willing to actively engage and help the users of
their code make it better, then how are the users suppose to help ??

While today I'm not a developer, I used to be a C and ASM coder, so writing
code is not a foreign thing.
The documentation on alerts (and many other parts) is non-existent.
What you are asking me (and others) to do, is dig through YOUR code and
figure out YOUR logic and how things work.
Then create documentation on that.
Not going to happen.  I'm trying to run a network, not develop software.

If on the other hand, you as developers cared, and actually submitted
enough usable data that the rest of us could work with,
then I suspect you would see more examples.

Heck, even after I made a call for examples, the number of examples I got
was ZERO.  Nothing from developers
Nothing from other users, NOTHING.   That tells me nobody cares.

My biggest frustration with LibreNMS and to a great extent Observium, is
that their isn't a good support model for
actual admins.  Going up to an IRC chat group, and MAYBE getting on the
right time zone and getting a usable response, doesn't work.
There is no functional email forum, like most other projects at scale have
and use.

We've moved to Zabbix for critical monitoring and functionality. We will
continue to use LibreNMS for a bit, but suspect we will migrate fully to
Zabbix.

We have a laundry list of issues with Libre, but given the high level of
friction in getting things adjusted, we've stopped maintaining a bug /
problem list.

There are lots of plusses for LibreNMS, but right now the negatives out
perform the positives.

**Comment by davidlanline** [comment]
interesting...I did apply it to the template, and most of my template works. 

only the "notes" and "purpose" fields come back blank, even though there are values filled in to both fields for the device that is alerting. 

***Update***

I tested with another host and everything came back like expected. 

Apparently there is something with that particular device which is not returning right, perhaps because the device has been down for a while? 

Perhaps if the device came back up, rediscovered and then went down again it would work? 


@digitalpilotnm 

I don't entirely disagree with you. Personally I like a lot of what I see in Libre, I think they are most of the way to the important features that I need, the one thing which is seriously lacking is the documentation. 

As far as that goes, I am continuing to work on getting my templates the way I like them and will offer you this for a device up/down template: 


```
<i>%hostname </i><br />
Severity: <b>%severity</b><br />


{if %faults}<b>Faults:</b><br />
{foreach %faults}<br />
</i>
<b>Reason Down:</b>%value.status_reason<br />
<b>Last Time Up:</b> %value.last_ping<br />
<b>Time Elapsed:</b> %elapsed <br />
<b>Device Description:</b>%value.sysDescr<br />
<b>Device Location:</b> %value.location<br />
<b>Device IP/AdminInfo:</b> %value.purpose<br />
<b>Device Notes:</b>%value.notes <br />
<b>Name:</b>%value.sysName<br />
<b>OS: </b>%value.os<br />
{/foreach}<br />
{/if}
To stop this alert acknowledge the alert here: <a href=http://myhostname.mydomain.com/alerts/> Monitoring System</a>
```

**Comment by crcro** [comment]
If you have a list of bugs you could submit them as everybody else, otherwise your are just a "kid" with a list.
If you want something that does things for you (like monitoring YOUR network) then you could give back at least a good talk/pacience. You want something that works after YOUR logic? code it yourself or go and pay someone to do it.

**Comment by laf** [comment]
@digitalpilotnm The only difference between a developer here and a user is that a developer is taking timeout to contribute something back. I'm no different than the next person who comes along to use LibreNMS except that I give back to the project and do so in many ways. To say that we aren't willing to actively engage or help people is a MASSIVE slap in the face - you only have to look at the commits here, the irc channel and even lately the community.librenms.org site. If people don't get back to you it's because they don't either have the time or know the answer.

What's really frustrating with this is the amount of people who do just read the docs and manage to get things up and running. They do contain information and whilst they may not be perfect they get the job done. That's not to say they can't be improved, they can but it's down to people having the time and as mentioned before, this is an open source project and not a business. You can read that as an excuse if you like.

@davidlanline Glad it's working for you. With regards to the one host that's not, a fresh alert shouldn't make a difference - data from the tables is pulled in realtime when the alert is sent so it's not cached in anyway. Do you have any strange characters in these fields which might be causing the template parsing to break, things like { } could cause problems but I've not tested to be sure.

**Comment by davidlanline** [comment]
@laf Actually the fresh alert seems to have resolved it. In doing some other testing this morning I changed the device down alert rule so it thought my original device was up. I then changed the rule back the device went down as it should and when I got the alert the notes and purpose fields were showing up properly. I have no idea why this would be, but it is the way it worked out. 

I appreciate the help. Perhaps if I can ask you some more clarifying questions about Templates specifically I can attempt to re-write/clraify the template section of the documentation for future users, I am beginning to understand the templates, but there are still definitely some pieces of the templates I do not understand.
